### PR TITLE
Agent V2

### PR DESCRIPTION
**Pull request by motiaFB**

### CODE REVIEW

File: api/customer-server.api.yml [hunk 1]
```diff
@@ -49,13 +49,13 @@ components:
...
-        - msgIds
+        - requestsIds
       properties:
-        msgIds:
+        requestsIds:
           type: array
           items:
-            type: number
-            example: 425878000014
+            type: string
```
Comment (nirfireblocks) [inline api/customer-server.api.yml:57]: how about adding "format: uuid" ?

File: api/customer-server.api.yml [hunk 2]
```diff
@@ -75,101 +75,203 @@ components:
...
-          type: string
-          enum: [PENDING_SIGN, SIGNED, FAILED]
-          example: SIGNED
-        errorMessage:
+          format: uuid
+        type:
+          $ref: "#/components/schemas/RequestType"
+    Message:
+      required:
+        - payload
```
Comment (nirfireblocks) [inline api/customer-server.api.yml:102]: Don't we want `payloadSignatureData` required as well?

File: api/customer-server.api.yml [hunk 3]
```diff
@@ -75,101 +75,203 @@ components:
...
+          type: string
+          format: uuid
+          example: 8c2b2b3d-fb83-497e-8138-72446b9184b6
         timestamp:
           type: number
           example: 1704122262
         version:
+          type: string
+          example: 1.0.0
+        metadata:
```
Comment (nirfireblocks) [inline api/customer-server.api.yml:171]: I'm confused between `metadata` and `TxMetadata`

File: api/customer-server.api.yml [hunk 4]
```diff
@@ -75,101 +75,203 @@ components:
...
+        type:
+          $ref: "#/components/schemas/ResponseType"
+        status:
           type: string
-          format: uuid
-          example: 70721651-a7f3-42f6-a984-6e058269495f
+          enum: [PENDING_SIGN, SIGNED, FAILED]
+          example: SIGNED
         requestId:
+          description: The unique request ID used exclusively for Proof of Ownership requests.
```
Comment (nirfireblocks) [inline api/customer-server.api.yml:230]: The requestId is used for all requests, not only Proof Of Ownership

File: examples/server/Dockerfile [hunk 5]
```diff
@@ -62,4 +62,4 @@ USER root
 
 EXPOSE 5000
 # CMD npm run start
-CMD npm run start:dev
+CMD ["npm", "run", "start:dev"]
```
Comment (nirfireblocks) [inline examples/server/Dockerfile:65]: Why is this needed?

File: examples/server/src/dao/messages.dao.ts [hunk 6]
```diff
@@ -18,41 +23,58 @@ const getMessagesCollection = async () => {
...
-      _id: msgId,
-      msgId,
-      requestId: message.requestId,
-      type,
-      message,
-      payload,
+      _id: transportMetadata.requestId,
+      type: newType,
       status: 'PENDING_SIGN',
+      requestId: transportMetadata.requestId,
```
Comment (nirfireblocks) [inline examples/server/src/dao/messages.dao.ts:47]: why do we need both `_id` and `requestId`?

File: examples/server/src/dao/messages.dao.ts [hunk 7]
```diff
@@ -18,41 +23,58 @@ const getMessagesCollection = async () => {
...
+
     return {
-      _id: msgId,
-      msgId,
-      requestId: message.requestId,
-      type,
-      message,
-      payload,
+      _id: transportMetadata.requestId,
+      type: newType,
```
Comment (nirfireblocks) [inline examples/server/src/dao/messages.dao.ts:45]: why are we storing the new type in cache? is it for matching the incoming response?

File: api/customer-server.api.yml (hunk 3 above)
Comment (motiaFB) [reply api/customer-server.api.yml:171]: metadata is a field name in MessagePayload. TxMetadata is a new type of the metadata field in KEY_LINK_TX_SIGN_REQUEST message payload

File: examples/server/Dockerfile (hunk 5 above)
Comment (motiaFB) [reply examples/server/Dockerfile:65]: Docker recommends this JSON format
https://docs.docker.com/reference/build-checks/json-args-recommended/

File: examples/server/src/dao/messages.dao.ts (hunk 7 above)
Comment (motiaFB) [reply examples/server/src/dao/messages.dao.ts:45]: It's not the cache, it's the example customer server DB. The DB row contains the response so the type should be the response type

File: src/customer-server-client.ts [hunk 8]
```diff
@@ -1,14 +1,23 @@
...
+        const { statuses: serverStatuses } = await customerServerApi.messagesStatus({ requestsIds });
+        await messagesService.updateStatus(serverStatuses.map((messagesStatus): ExtendedMessageStatus => {
+          const decodedMsg = messages.find((msg) => msg.messageStatus.requestId === messagesStatus.requestId);
+          return {
+            msgId: decodedMsg.msgId,
+            request: decodedMsg.request,
+            messageStatus: messagesStatus,
+          };
+        }));
       }
```
Comment (nirfireblocks) [inline src/customer-server-client.ts:21]: I'm confused by this.
I'm not sure what is updated, and also realized we have two cachees - one for the messages and one for the msgIds

Comment (motiaFB) [reply src/customer-server-client.ts:21]: The `updateStatus` function expects `ExtendedMessageStatus` (which is the cache value type) so this `map` merges the cache values (request and msgId) and the new customer-server statuses

File: examples/server/src/services/hsm-sign-service.ts [hunk 9]
```diff
@@ -1,61 +1,77 @@
...
       }
-
-      const { signingDeviceKeyId, data } = msg.message;
-      if (msg.status === 'SIGNED') {
-        msg.signedPayload = await hsmFacade.sign(signingDeviceKeyId, data, algorithm);
+      else {
+        const { signingDeviceKeyId, messagesToSign } = msg.message;
+        for (const { message, index } of messagesToSign) {
+          const signature = await hsmFacade.sign(signingDeviceKeyId, message, algorithm);
+          msg.response.signedMessages.push({ index, signature, message });
```
[the remark this answers is not in the document]
Comment (motiaFB) [reply examples/server/src/services/hsm-sign-service.ts:30]: done

File: examples/server/src/dao/messages.dao.ts (hunk 6 above)
Comment (motiaFB) [reply examples/server/src/dao/messages.dao.ts:47]: `_id` is the DB primary field, `requestId` is a field part of `MessageStatus` struct.
I split the `MessageStatus` fields initialization and `DbMsg` fields initialization

File: api/customer-server.api.yml (hunk 4 above)
Comment (motiaFB) [reply api/customer-server.api.yml:230]: done

File: api/customer-server.api.yml (hunk 2 above)
Comment (motiaFB) [reply api/customer-server.api.yml:105]: done

File: api/customer-server.api.yml (hunk 1 above)
Comment (motiaFB) [reply api/customer-server.api.yml:58]: done

File: src/services/fb-server.api.ts [hunk 10]
```diff
@@ -41,27 +42,30 @@ const fbServerApi = {
...
+      const responseObject = {
+        type,
+        status: msgStatus.status,
+        request: parsedPayload,
+        response: msgStatus.response,
       };
-      const url = `${MOBILE_GATEWAY_URL}/${TYPE_TO_ENDPOINT[responseType]}`;
-      logger.info(`broadcasting to ${url} response ${JSON.stringify(requestObject)}`);
+
+      if (!TYPE_TO_ENDPOINT.hasOwnProperty(type)) {
```
Comment (nirfireblocks) [inline src/services/fb-server.api.ts:59]: Of all ways to check this `hasOwnProperty` sounds wrong, why not `in`?
https://www.geeksforgeeks.org/how-to-check-if-a-key-exists-in-a-dictionary-in-typescript/

File: src/services/messages.service.ts [hunk 11]
```diff
@@ -1,55 +1,127 @@
...
+  getPendingMessages(): ExtendedMessageStatus[];
   handleMessages(messages: FBMessageEnvelope[]): Promise<void>;
-  updateStatus(messagesStatus: MessageStatus[]): Promise<void>;
+  updateStatus(messagesStatus: ExtendedMessageStatus[]): Promise<void>;
 }
 
 class MessageService implements IMessageService {
-  private msgCache: { [msgId: number]: MessageStatus } = {};
-  private knownMessageTypes: TxType[] = ['EXTERNAL_KEY_PROOF_OF_OWNERSHIP_REQUEST', 'EXTERNAL_KEY_SIGNING_REQUEST'];
+  private MAX_CACHE_SIZE = 1024;
```
Comment (nirfireblocks) [inline src/services/messages.service.ts:14]: I think we can leave it for now, but in the future I think it should be a parameter in the .env file (and 1024 be a default value). I will also help with tests.

File: src/services/messages.service.ts [hunk 12]
```diff
@@ -1,55 +1,127 @@
...
+          this.msgCache[messageStatus.requestId].msgId = msgId;
         }
-        if (msgStatus.status === 'SIGNED' || msgStatus.status === 'FAILED') {
-          logger.info(`Got signed message id ${msgStatus.msgId}`);
-          await fbServerApi.broadcastResponse(msgStatus);
-          await fbServerApi.ackMessage(msgStatus.msgId);
-          delete this.msgCache[msgStatus.msgId];
+
+        if (status === 'SIGNED' || status === 'FAILED') {
+          logger.info(`Got signed message id ${msgId}, cacheId: ${requestId}`);
```
Comment (nirfireblocks) [inline src/services/messages.service.ts:108]: Different log for failed

File: src/services/messages.service.ts [hunk 13]
```diff
@@ -1,55 +1,127 @@
...
+          this.msgCache[messageStatus.requestId].messageStatus = messageStatus;
+
+          await fbServerApi.broadcastResponse(messageStatus, request);
+          await fbServerApi.ackMessage(msgId);
         }
+      } catch (e) {
+        throw new Error(`Error updating status to fireblocks ${e.message}`);
```
Comment (nirfireblocks) [inline src/services/messages.service.ts:115]: What do you think of printing the `msgStatus` here?

File: src/utils/messages-utils.test.ts [hunk 14]
```diff
@@ -51,7 +54,115 @@ describe('Messages utils', () => {
...
+    const invalid_version = "0.0.0";
+    const fbMessage = aCustomFbProofOfOwnershipMessage(privateKey, { version: invalid_version });
+    const fbMessageEnvelope = buildASignedMessage(fbMessage, certificates.zs);
+
+    const expectToThrow = () => utils.decodeAndVerifyMessage(fbMessageEnvelope, certificates);
+
+    expect(expectToThrow).toThrowErrorMatchingInlineSnapshot(`"Unsupported message version: ${invalid_version}"`);
+  });
+
+  it('should verify unknown message', () => {
```
Comment (nirfireblocks) [inline src/utils/messages-utils.test.ts:104]: Why should we verify this?

File: src/utils/messages-utils.test.ts [hunk 15]
```diff
@@ -51,7 +54,115 @@ describe('Messages utils', () => {
...
+      message: fbMessage.payload,
+      transportMetadata: {
+        requestId: txId,
+        type: fbMessage.type,
+      },
+    };
+    expect(messageEnvelope).toEqual({ request: expectedMessage, msgId: fbMessageEnvelope.msgId });
+  });
+
+  it('should not verify a tx sign request message with false ps signature', () => {
```
Comment (nirfireblocks) [inline src/utils/messages-utils.test.ts:151]: Why only ps? why not version here as well?
Not saying we should add all, just wondering.

File: src/utils/messages-utils.ts [hunk 16]
```diff
@@ -66,31 +63,83 @@ function verifyRSASignatureFromCertificate(
...
+      signature: payloadSignature.signature,
+      format: 'hex',
+    },
+  };
+}
+
+const getPayloadVerifyDetails = (fbMessage: FBMessage): VerifyDetails => {
+  const fbMsgPayload = fbMessage.payload;
+  const payloadSignatureData = fbMsgPayload.payloadSignatureData;
+  if (payloadSignatureData === undefined && payloadSignatureData === null) {
```
Comment (nirfireblocks) [inline src/utils/messages-utils.ts:91]: I think this should be `||` and not `&&`

File: api/customer-server.api.yml (hunk 3 above)
Comment (nirfireblocks) [reply api/customer-server.api.yml:174]: Where does TxMetadata come from, core?

Comment (motiaFB) [reply api/customer-server.api.yml:174]: yes, policy-service

File: src/utils/messages-utils.ts (hunk 16 above)
Comment (motiaFB) [reply src/utils/messages-utils.ts:91]: done

File: src/services/messages.service.ts (hunk 13 above)
Comment (motiaFB) [reply src/services/messages.service.ts:115]: done

File: src/services/messages.service.ts [hunk 17]
```diff
@@ -1,55 +1,127 @@
...
   handleMessages(messages: FBMessageEnvelope[]): Promise<void>;
-  updateStatus(messagesStatus: MessageStatus[]): Promise<void>;
+  updateStatus(messagesStatus: ExtendedMessageStatus[]): Promise<void>;
 }
 
 class MessageService implements IMessageService {
-  private msgCache: { [msgId: number]: MessageStatus } = {};
-  private knownMessageTypes: TxType[] = ['EXTERNAL_KEY_PROOF_OF_OWNERSHIP_REQUEST', 'EXTERNAL_KEY_SIGNING_REQUEST'];
+  private MAX_CACHE_SIZE = 1024;
+  private msgCache: { [requestId: string]: ExtendedMessageStatus } = {};
```
[the remark this answers is not in the document]
Comment (motiaFB) [reply src/services/messages.service.ts:15]: done

File: src/services/messages.service.ts (hunk 11 above)
Comment (motiaFB) [reply src/services/messages.service.ts:14]: done

File: src/services/fb-server.api.ts (hunk 10 above)
Comment (motiaFB) [reply src/services/fb-server.api.ts:59]: done
